vault/token_store.go

```diff
@@ -917,23 +1069,187 @@ func (ts *TokenStore) destroyCubbyhole(saltedID string) error {
 
 func (ts *TokenStore) authRenew(
 	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	if req.Auth == nil {
+		return nil, fmt.Errorf("request auth is nil")
+	}
+
+	f := framework.LeaseExtend(req.Auth.Increment, 0, ts.System())
+
+	idInt, ok := req.Auth.InternalData["id"]
+	if !ok {
+		// Fall back here; this is pre-roles so there are no stored IDs, so use previous behavior
+		return f(req, d)
+	}
+
+	id, ok := idInt.(string)
+	if !ok {
+		return nil, fmt.Errorf("found id in internal data but could not interpret as string")
+	}
+
+	te, err := ts.Lookup(id)
+	if err != nil {
+		return nil, fmt.Errorf("error looking up token: %s", err)
+	}
+	if te == nil {
+		return nil, fmt.Errorf("no token entry found during lookup")
+	}
+
+	// No role? Use normal LeaseExtend semantics
+	if te.Role == "" {
+		return f(req, d)
+	}
+
+	role, err := ts.tokenStoreRole(te.Role)
+	if err != nil {
+		return nil, fmt.Errorf("error looking up role %s: %s", te.Role, err)
+	}
+
+	if role == nil {
+		resp, err := f(req, d)
```

Comment:
That was my original thought. I was worried about the case where a role is deleted but the maximum TTL of the token is not reached, but I guess better to simply not renew and let someone deal with the error. If they're renewing early enough it should be okay.